shoop/core/models/_orders.py

```diff
@@ -514,6 +515,100 @@ def create_shipment(self, product_quantities, supplier=None, shipment=None):
+        `OrderLineType.PRODUCT` and the `restock_products` boolean
+        flag is set to `True`, the products will be restocked with the
+        order's supplier the exact amount of the value of the `quantity`
+        field.
+
+        :param refund_data: List of dicts containing refund data.
+        :type refund_data: [dict]
+        :param created_by: Refund creator's user instance, used for
+                           adjusting supplier stock.
+        :type created_by: django.contrib.auth.User|None
+        """
+        index = self.lines.order_by("ordering").last().ordering
+        refund_lines = []
+        for refund in refund_data:
+            index += 1
+            amount_value = refund.get("amount", 0)
+            quantity = refund.get("quantity", 0)
+            parent_line = refund.get("line")
+            restock_products = refund.get("restock_products")
+
+            # TODO: Also raise this if the sum value of refunds exceeds,
+            #       and do so before creating any order lines
+            self.cache_prices()
+            if amount_value > self.taxful_total_price.amount.value:
+                raise RefundExceedsAmountException
+
+            unit_price = parent_line.base_unit_price.amount.value if parent_line else 0
+            total_price = unit_price * quantity + amount_value
+
+            refund_line = OrderLine.objects.create(
+                text=_("Refund for %s" % parent_line.text) if parent_line else _("Manual refund"),
+                order=self,
+                type=OrderLineType.REFUND,
+                parent_line=parent_line,
+                ordering=index,
+                base_unit_price_value=Money(total_price * (-1), self.currency),
+                quantity=1
+            )
+            refund_lines.append(refund_line)
+
+            if parent_line and parent_line.type == OrderLineType.PRODUCT:
+                product = parent_line.product
+            else:
+                product = None
+
+            if restock_products and quantity and product and (product.stock_behavior == StockBehavior.STOCKED):
+                parent_line.supplier.adjust_stock(product.id, quantity, created_by=created_by)
+
+        self.cache_prices()
+        self.save()
+        refund_created.send(sender=type(self), order=self, refund_lines=refund_lines)
+
+    def create_full_refund(self, restock_products=False):
+        """
+        Create a full for entire order contents, with the option of
+        restocking stocked products.
+
+        :param restock_products: Boolean indicating whether to restock products
+        :type restock_products: bool|False
+        """
+        if self.has_refunds():
+            raise NoRefundToCreateException
+        self.cache_prices()
+        amount = self.taxful_total_price
+        self.create_refund([{"amount": amount.value}])
+        if restock_products:
+            for product_line in self.lines.filter(type=OrderLineType.PRODUCT):
```

Comment:
Could this work: `for product_line in self.lines.filter(type=OrderLineType.PRODUCT, product__stock_behavior=StockBehavior.STOCKED)`